File: src/main/java/com/mystery2099/util/VoxelShapeHelper.java
package com.mystery2099.util;

import net.minecraft.util.function.BooleanBiFunction;
import net.minecraft.util.math.Direction;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;

import java.util.Collection;
import java.util.concurrent.atomic.AtomicReference;

public class VoxelShapeHelper {
    public static VoxelShape combineAll(Collection<VoxelShape> shapes) {
        return shapes.stream().reduce(VoxelShapes::union).orElse(VoxelShapes.empty());
    }

    public static VoxelShape setMaxHeight(VoxelShape source, double height) {
        var result = new AtomicReference<>(VoxelShapes.empty());
        source.forEachBox((minX, minY, minZ, maxX, maxY, maxZ) -> {
            var shape = VoxelShapes.cuboid(minX, minY, minZ, maxX, height, maxZ);
            result.set(VoxelShapes.combine(result.get(), shape, BooleanBiFunction.OR));
        });
        return result.get().simplify();
    }

    public static VoxelShape limitHorizontal(VoxelShape source) {
        var result = new AtomicReference<>(VoxelShapes.empty());
        source.forEachBox((minX, minY, minZ, maxX, maxY, maxZ) -> {
            var shape = VoxelShapes.cuboid(limit(minX), minY, limit(minZ), limit(maxX), maxY, limit(maxZ));
            result.set(VoxelShapes.combine(result.get(), shape, BooleanBiFunction.OR));
        });
        return result.get().simplify();
    }

    public static VoxelShape rotate(VoxelShape source, Direction direction) {
        var adjustedValues = adjustValues(direction, source.getMin(Direction.Axis.X), source.getMin(Direction.Axis.Z), source.getMax(Direction.Axis.X), source.getMax(Direction.Axis.Z));
        return VoxelShapes.cuboid(adjustedValues[0], source.getMin(Direction.Axis.Y), adjustedValues[1], adjustedValues[2], source.getMax(Direction.Axis.Y), adjustedValues[3]);
    }
    public static VoxelShape rotate(Collection<VoxelShape> sources, Direction direction) {
        var output = new AtomicReference<>(VoxelShapes.empty());
        sources.forEach(shape -> output.set(VoxelShapes.union(output.get(), rotate(shape, direction))));
        return output.get();
    }

    private static double[] adjustValues(Direction direction, double var1, double var2, double var3, double var4) {
        return switch (direction) {
            //forward
            case WEST -> new double[]{1.0F - var3, 1.0F - var4, 1.0F - var1, 1.0F - var2};
            //right
            case NORTH -> new double[]{var2, 1.0F - var3, var4, 1.0F - var1};
            //left
            case SOUTH -> new double[]{1.0F - var4, var1, 1.0F - var2, var3};
            default -> new double[]{var1, var2, var3, var4};
        };
    }

    private static double limit(double value) {
        return Math.max(0.0, Math.min(1.0, value));
    }
}

File: src/main/kotlin/com/mystery2099/util/VoxelShapeHelper.kt
package com.mystery2099.util

import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import java.util.concurrent.atomic.AtomicReference
import kotlin.math.max
import kotlin.math.min

object VoxelShapeHelper {
    private val Double.limited: Double
        get() = this.limit()
    val Collection<VoxelShape>.combined: VoxelShape
        get() = this.union()
    val Array<VoxelShape>.combined: VoxelShape
        get() = this.union()

    fun Collection<VoxelShape>.union(): VoxelShape = this.fold(VoxelShapes.empty(), VoxelShapes::union)
    fun Array<VoxelShape>.union(): VoxelShape = this.fold(VoxelShapes.empty(), VoxelShapes::union)

    fun VoxelShape.unionWith(otherShape: VoxelShape): VoxelShape = VoxelShapes.union(this, otherShape)
    fun VoxelShape.unionWith(vararg otherShapes: VoxelShape): VoxelShape = otherShapes.fold(this, VoxelShapes::union)
    fun VoxelShape.unionWith(otherShapes: Collection<VoxelShape>): VoxelShape = this.unionWith(otherShapes.combined)

    fun setMaxHeight(source: VoxelShape, height: Double): VoxelShape {
        val result = AtomicReference(VoxelShapes.empty())
        source.forEachBox { minX: Double, minY: Double, minZ: Double, maxX: Double, maxY: Double, maxZ: Double ->
            val shape = VoxelShapes.cuboid(minX, minY, minZ, maxX, height, maxZ)
            result.set(result.get().unionWith(shape))
        }
        return result.get()
    }

    fun limitHorizontal(source: VoxelShape): VoxelShape {
        val result = AtomicReference(VoxelShapes.empty())
        source.forEachBox { minX: Double, minY: Double, minZ: Double, maxX: Double, maxY: Double, maxZ: Double ->
            val shape = VoxelShapes.cuboid(minX.limited, minY, minZ.limited, maxX.limited, maxY, maxZ.limited)
            result.set(result.get().unionWith(shape))
        }
        return result.get()
    }

    //Do not use on complex VoxelShapes
    fun VoxelShape.rotate(direction: VoxelShapeRotations): VoxelShape {
        val adjustedValues = adjustValues(
            direction, this.getMin(Direction.Axis.X), this.getMin(Direction.Axis.Z), this.getMax(
                Direction.Axis.X
            ), this.getMax(Direction.Axis.Z)
        )
        return VoxelShapes.cuboid(
            adjustedValues[0], this.getMin(Direction.Axis.Y),
            adjustedValues[1], adjustedValues[2], this.getMax(Direction.Axis.Y), adjustedValues[3]
        )
    }

    //Do not use on complex VoxelShapes
    fun VoxelShape.rotateLeft(): VoxelShape = rotate(VoxelShapeRotations.LEFT)

    //Do not use on complex VoxelShapes
    fun VoxelShape.rotateRight(): VoxelShape = rotate(VoxelShapeRotations.RIGHT)

    //Do not use on complex VoxelShapes
    fun VoxelShape.flip(): VoxelShape = rotate(VoxelShapeRotations.FLIP)

    fun Collection<VoxelShape>.rotateElements(direction: VoxelShapeRotations): Collection<VoxelShape> {
        if (isEmpty()) {
            val callerStackTrace = Thread.currentThread().stackTrace[2]
            val callerClassName = callerStackTrace.className
            val callerLineNumber = callerStackTrace.lineNumber
            println("Warning: Collection of VoxelShapes is empty in class: $callerClassName, line: $callerLineNumber. Returning a list containing an empty VoxelShape.")
            return listOf(VoxelShapes.empty())
        }
        return map { it.rotate(direction) }
    }
    fun Array<VoxelShape>.rotateElements(direction: VoxelShapeRotations): Array<VoxelShape> {
        if (isEmpty()) {
            val callerStackTrace = Thread.currentThread().stackTrace[2]
            val callerClassName = callerStackTrace.className
            val callerLineNumber = callerStackTrace.lineNumber
            println("Warning: Array of VoxelShapes is empty in class: $callerClassName, line: $callerLineNumber. Returning an Array containing an empty VoxelShape.")
            return arrayOf(VoxelShapes.empty())
        }
        return map { it.rotate(direction) }.toTypedArray()
    }

    fun Collection<VoxelShape>.rotateAndCombine(direction: VoxelShapeRotations): VoxelShape {
        return rotateElements(direction).combined
    }
    fun Array<VoxelShape>.rotateAndCombine(direction: VoxelShapeRotations): VoxelShape {
        return rotateElements(direction).combined
    }

    private fun adjustValues(
        direction: VoxelShapeRotations,
        var1: Double,
        var2: Double,
        var3: Double,
        var4: Double
    ): DoubleArray = when (direction) {
        VoxelShapeRotations.FLIP -> doubleArrayOf(1.0f - var3, 1.0f - var4, 1.0f - var1, 1.0f - var2)
        VoxelShapeRotations.RIGHT -> doubleArrayOf(var2, 1.0f - var3, var4, 1.0f - var1)
        VoxelShapeRotations.LEFT -> doubleArrayOf(1.0f - var4, var1, 1.0f - var2, var3)
        else -> doubleArrayOf(var1, var2, var3, var4)
    }

    private fun Double.limit(): Double = max(0.0, min(1.0, this))

}
enum class VoxelShapeRotations {
    LEFT, RIGHT, FLIP, UP, DOWN
}